hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go

```diff
@@ -41,266 +27,5 @@ func SetupWebhookWithManager(mgr ctrl.Manager) error {
 	if err != nil {
 		return fmt.Errorf("unable to register hostedcontrolplane webhook: %w", err)
 	}
-	err = ctrl.NewWebhookManagedBy(mgr).
-		For(&hyperv1.AWSEndpointService{}).
-		Complete()
-	if err != nil {
-		return fmt.Errorf("unable to register awsendpointservice webhook: %w", err)
-	}
-	return nil
-
-}
-
-var _ webhook.CustomValidator = &Webhook{}
-
-// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type.
-func (webhook *Webhook) ValidateCreate(ctx context.Context, obj runtime.Object) error {
-	hostedCluster, ok := obj.(*hyperv1.HostedCluster)
-	if !ok {
-		return apierrors.NewBadRequest(fmt.Sprintf("expected a HostedCluster but got a %T", obj))
-	}
-
-	return validateHostedClusterCreate(hostedCluster)
-}
-
-type cidrEntry struct {
-	net  net.IPNet
-	path field.Path
-}
-
-func cidrsOverlap(net1 *net.IPNet, net2 *net.IPNet) error {
-	if net1.Contains(net2.IP) || net2.Contains(net1.IP) {
-		return fmt.Errorf("%s and %s", net1.String(), net2.String())
-	}
 	return nil
 }
-
-func compareCIDREntries(ce []cidrEntry) field.ErrorList {
-	var errs field.ErrorList
-
-	for o := range ce {
-		for i := o + 1; i < len(ce); i++ {
-			if err := cidrsOverlap(&ce[o].net, &ce[i].net); err != nil {
-				errs = append(errs, field.Invalid(&ce[o].path, ce[o].net.String(), fmt.Sprintf("%s and %s overlap: %s", ce[o].path.String(), ce[i].path.String(), err)))
-			}
-		}
-	}
-	return errs
-}
-
-func validateSliceNetworkCIDRs(hc *hyperv1.HostedCluster) field.ErrorList {
-	var cidrEntries []cidrEntry
-
-	for _, cidr := range hc.Spec.Networking.MachineNetwork {
-		ce := cidrEntry{(net.IPNet)(cidr.CIDR), *field.NewPath("spec.networking.MachineNetwork")}
-		cidrEntries = append(cidrEntries, ce)
-	}
-	for _, cidr := range hc.Spec.Networking.ServiceNetwork {
-		ce := cidrEntry{(net.IPNet)(cidr.CIDR), *field.NewPath("spec.networking.ServiceNetwork")}
-		cidrEntries = append(cidrEntries, ce)
-	}
-	for _, cidr := range hc.Spec.Networking.ClusterNetwork {
-		ce := cidrEntry{(net.IPNet)(cidr.CIDR), *field.NewPath("spec.networking.ClusterNetwork")}
-		cidrEntries = append(cidrEntries, ce)
-	}
-
-	return compareCIDREntries(cidrEntries)
-}
-
-func validateKubevirtBaseDomainPassthroughCreate(hc *hyperv1.HostedCluster) *field.Error {
-
-	// It is invalid for someone to enable the BaseDomainPassthrough feature
-	// and attempt to set their own custom BaseDomain during HC Creation.
-	//
-	// The BaseDomainPassthrough feature autocreates the BaseDomain for the user.
-	if hc.Spec.Platform.Type == hyperv1.KubevirtPlatform &&
-		hc.Spec.Platform.Kubevirt != nil &&
-		hc.Spec.Platform.Kubevirt.BaseDomainPassthrough != nil &&
-		*hc.Spec.Platform.Kubevirt.BaseDomainPassthrough &&
-		hc.Spec.DNS.BaseDomain != "" {
-
-		return field.InternalError(
-			field.NewPath("HostedCluster.spec.platform.kubevirt.baseDomainPassthrough"),
-			errors.New("BaseDomain can not be set when KubeVirt's BaseDomainPassthrough feature is enabled"))
-	}
```

Comment:
how do we express something like this (where one tuning influences another tuning) without the validation webhook?

Reply:
Couldn't this be expressed via CEL rule?
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
https://github.com/google/cel-spec/blob/v0.6.0/doc/langdef.md#evaluation

e.g 
```
has(platform.kubevirt.baseDomainPassthrough) && !has(self.BaseDomain) || !has(platform.kubevirt.baseDomainPassthrough) && has(self.BaseDomain)```

@deads2k @JoelSpeed is there a best practices docs to express the above, i.e. fields discrimination?

@davidvossel I'd also expect us to check this in the backend and fail in a condition.

Reply:
It looks like you could use a CEL validation here as Alberto has stated

@enxebre What exactly are you looking for? We have some guidance on discriminated unions, https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions, though I'd expect, looking at `platform.kubevirt` that you already have a DU in place for platform specific options.

Given `BaseDomain` is outside of this DU (and actually in a completely different struct) it's a bit awkward because the fields are "far apart", but I would certainly recommend having a godoc string on each of the two fields explaining that they are mutually exclusive. 

```
// The usage of this field requires that `spec.dns.baseDomain` is omitted.
BaseDomainPassThrough ...
```
```
// When using the KubeVirt platform, if the `baseDomainPassthrough` feature is ..., this field must be omitted.
BaseDomain ...
```